### PR TITLE
🐛 Use RepoClient.Search API in SAST check

### PR DESCRIPTION
**Pull request by azeemshaikh38**
* **Please check if the PR fulfills these requirements**
- [ ] Tests for the changes have been added (for bug fixes / features)
- [ ] PR title follows the guidelines defined in  https://github.com/ossf/scorecard/blob/main/CONTRIBUTING.md#pr-process

* **What kind of change does this PR introduce?** (Bug fix, feature, docs update, ...)
Update SAST check to use the new RepoClient.Search API.

* **Does this PR introduce a breaking change?** (What changes might users need to make in their application due to this PR?)
No.

### CODE REVIEW

File: clients/githubrepo/search.go [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+)
+
+type searchHandler struct {
+	ghClient *github.Client
+	ctx      context.Context
+	owner    string
+	repo     string
+}
+
+func (handler *searchHandler) init(ctx context.Context, owner, repo string) error {
```
Comment (naveensrinivasan) [inline clients/githubrepo/search.go:35]: Why should this func return an error?

File: clients/githubrepo/search.go [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+	resp, _, err := handler.ghClient.Search.Code(handler.ctx, query, &github.SearchOptions{})
+	if err != nil {
+		return clients.SearchResponse{}, fmt.Errorf("Search.Code: %w", err)
+	}
+	return searchResponseFrom(resp), nil
+}
+
+func (handler *searchHandler) buildQuery(request clients.SearchRequest) (string, error) {
+	if request.Query == "" {
+		// nolint: goerr113
```
Comment (naveensrinivasan) [inline clients/githubrepo/search.go:57]: Why not address the lint issue instead of `nolint`?

Comment (inferno-chromium) [reply clients/githubrepo/search.go:57]: +1, i see quite a lot of goerr113 in other places, should we just disable this linter altogether or fix these ?

File: clients/githubrepo/search.go [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+		return "", errors.New("search query is empty")
+	}
+	var queryBuilder strings.Builder
+	if _, err := queryBuilder.WriteString(
```
Comment (inferno-chromium) [inline clients/githubrepo/search.go:61]: Why do three api calls to queryBuilder.WriteString if the request.Filename and request.Path are set. so i think better to build the string first and then do just one queryBuilder.WriteString call.

**Review by inferno-chromium — APPROVED** [review]
Thanks! LGTM with few fixes needed.

File: clients/githubrepo/search.go (hunk 2 above)
Comment (azeemshaikh38) [reply clients/githubrepo/search.go]: Done.

File: clients/githubrepo/search.go [hunk 4]
```diff
@@ -0,0 +1,88 @@
...
+	handler.ctx = ctx
+	handler.owner = owner
+	handler.repo = repo
+	return nil
+}
+
+func (handler *searchHandler) search(request clients.SearchRequest) (clients.SearchResponse, error) {
+	query, err := handler.buildQuery(request)
+	if err != nil {
+		return clients.SearchResponse{}, fmt.Errorf("%w", err)
```
[the remark this answers is not in the document]
Comment (azeemshaikh38) [reply clients/githubrepo/search.go]: Done.

File: clients/githubrepo/search.go (hunk 1 above)
Comment (azeemshaikh38) [reply clients/githubrepo/search.go]: Done.

File: clients/githubrepo/search.go (hunk 3 above)
Comment (azeemshaikh38) [reply clients/githubrepo/search.go:61]: Not sure how you mean `build the string first`. `queryBuilder.WriteString` is just an optimized way of concatenating strings in Golang.

Comment (naveensrinivasan) [reply clients/githubrepo/search.go:61]: +1 with @azeemshaikh38 on this.

**Review by naveensrinivasan — APPROVED** [review]
Thanks!

File: clients/githubrepo/search.go (hunk 3 above)
Comment (inferno-chromium) [reply clients/githubrepo/search.go:61]: Ok, thanks for explaining, i got confused if this was making an actual github api call.

**Pull request merged**